### PR TITLE
amp-bind: Expression complexity limit

### PR DESCRIPTION
**Pull request by choumx**
Partial for #8057.

- Add complexity limit for BindExpression to 50 nodes in the AST. "50" is a magic number that's double the size of expressions used in `examples/bind/performance.amp.html`.
- Ignore this limit when testing manually, otherwise "game" and "turing" examples won't work. 😛

### CODE REVIEW

File: extensions/amp-bind/0.1/bind-expression.js [hunk 1]
```diff
@@ -85,19 +86,36 @@ const FUNCTION_WHITELIST = (function() {
...
 
     /** @const @private {!./bind-expr-defines.AstNode} */
     this.ast_ = parser.parse(this.expressionString);
+
+    // Check if this expression string is too large (for performance).
+    const size = this.numberOfNodesInAst_(this.ast_);
+    const maxSize = opt_maxAstSize || DEFAULT_MAX_AST_SIZE;
+    const skipConstraint = getMode().localDev && !getMode().test;
+    if (size > maxSize && !skipConstraint) {
+      throw new Error(`Expression size (${size}) exceeds max (${maxSize}): ` +
```
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-expression.js]: This is going to be the message printed in the console right? If so, I'm wondering if we should give more context here. Someone unfamiliar with the amp-bind implementation might not know what to make of this message.

File: extensions/amp-bind/0.1/bind-expression.js [hunk 2]
```diff
@@ -85,19 +86,36 @@ const FUNCTION_WHITELIST = (function() {
...
-  constructor(expressionString) {
+  constructor(expressionString, opt_maxAstSize) {
     /** @const {string} */
     this.expressionString = expressionString;
 
     /** @const @private {!./bind-expr-defines.AstNode} */
     this.ast_ = parser.parse(this.expressionString);
+
+    // Check if this expression string is too large (for performance).
+    const size = this.numberOfNodesInAst_(this.ast_);
```
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-expression.js]: Just curious: what's the performance impact of doing this?

File: extensions/amp-bind/0.1/bind-expression.js (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expression.js]: Actually this error will eventually be passed to `reportError()`, which highlights the element that contains this expression in dev console.

File: extensions/amp-bind/0.1/bind-expression.js (hunk 2 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-expression.js]: Good question. Looks like a few ms in total on `examples/bind/performance.amp.html` with 5x throttling.

File: extensions/amp-bind/0.1/bind-expression.js (hunk 1 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-expression.js]: Nonetheless, I'm not sure that the error gives a clear picture of the problem to the uninitiated. Maybe add something like "Expression contains too many literals, variables, or function calls". Up to you.

Comment (choumx) [inline extensions/amp-bind/0.1/bind-expression.js]: Oh I see, I thought you meant contextual location rather than content. Done!

**Comment by choumx** [comment]
Thanks for the reviews!

File: extensions/amp-bind/0.1/bind-expression.js [hunk 3]
```diff
@@ -85,19 +86,36 @@ const FUNCTION_WHITELIST = (function() {
...
+    const maxSize = opt_maxAstSize || DEFAULT_MAX_AST_SIZE;
+    const skipConstraint = getMode().localDev && !getMode().test;
+    if (size > maxSize && !skipConstraint) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-expression.js]: This needs a warning if we're in dev mode and we make it too big.

**Pull request merged**